# 方差分析

## 方差分析不需要真实的标签值

方差分析是用于两个及两个以上总体均值差别的显著性检验．例如，生产某种产品，为了使生产过程稳定，达到优质、高产，需要对影响产品质量的因素进行分析，找出有显著影响的那些因素，方差分析就是鉴别各因素效应的一种有效的统计方法.


### 1. 单因素方差分析（One-way analysis of variance）

若在一项试验中，只考虑一个因素 A 的变化，其他因素保待不变，称这种试验为单因素试验.具体做法是： A 取几个水平，在每个水平上作若干个试验，试验过程中除 A 外其他影响指标的因素都保持不变（只有随机因素存在），我们的任务是从试验结果推断，因素 A 对指标有无显著影响，即当 A 取不同水平时指标有无显著差别． A 取某个水平下的指标可视为随机变量，判断 A 取不同水平时指标有无显著差异，相当于检验若干总体的均值是否相等．

用 4 种工艺$A_1$, $A_2$, $A_3$, $A_4$生产灯泡，从各种工艺制成的灯泡中各抽出了若干个测量其寿命，结果如表 4.10 所示，试推断这几种工艺制成的灯泡寿命是否有显著差异.

\begin{center}
\begin{tabular}{c|cccc}
\hline
\multicolumn{5}{c}{表 4.10 4 种工艺生产的灯泡寿命\qquad\qquad(单位: h)} \\
\hline
$A_1$ & $A_2$ & $A_3$ & $A_4$ \\
\hline
1620 & 1580 & 1460 & 1500 \\
1670 & 1600 & 1540 & 1550 \\
1700 & 1640 & 1620 & 1610 \\
1750 & 1720 & 1680 &  \\
1800 &  &  &  \\
\hline
\end{tabular}
\end{center}

该例中不同配料的灯丝称为因素或因子，属单因素的试验，灯丝的不同配料方案称为水平，共有$A_1$, $A_2$, $A_3$, $A_4$四种水平. 一般情况下，单因素$A$的$r$种不同水平分别记为$A_1$, $A_2$, $\cdots$, $A_r$.

例 4.22 中在不同配料方案下各抽取若干样品作检验，每一种方案生产出灯泡的寿命都构成一个总体，为此要检验各种灯丝配料方案对灯泡寿命是否有显著影响，即检验四个总体的均值是否相等. 由于在实际问题中通常遇到的是各总体都服从或近似服从正态分布的情形，故可将这类问题归结为如下的数学模型.

设有$r$个正态总体$X_i \sim N(\mu_i, \sigma^2)$ ($i = 1, 2, \cdots, r$) 且相互独立. 对这$r$个总体作如下假设：
$$
H_0: \mu_1 = \mu_2 = \cdots = \mu_r.
$$

现独立地从各总体中随机抽取若干样品，如表 4.11 所列，要求利用该表中数据检验假设$H_0$是否成立.

\begin{center}
\begin{tabular}{c|c|c|c}
\hline
\multicolumn{4}{c}{表 4.11 样品数据表} \\
\hline
总体 & $X_1$ & $X_2$ & $X_r$ \\
\hline
样本 & $X_{11}, X_{12}, \cdots, X_{1n_1}$ & $X_{21}, X_{22}, \cdots, X_{2n_2}$ & $X_{r1}, X_{r2}, \cdots, X_{rn_r}$ \\
样本均值 & $\bar{X}_1$ & $\bar{X}_2$ & $\bar{X}_r$ \\
\hline
\end{tabular}
\end{center}

由假设检验的知识，对于假设$H_0$，可以用$t$检验法来检验任何两相邻总体均值是否相等就可以了，但这样就需要检验$r - 1$次，如果$r$较大时会很烦琐，为此采用离差分解法来分析. 将每个总体中抽出来的样品组成一组，共有$r$组，记各组内的样本平均值为
$$
\bar{X}_i = \frac{1}{n_i} \sum_{j=1}^{n_i} X_{ij},\quad i = 1, 2, \cdots, r,
$$
所以样本的总平均值为
$$
\bar{X} = \frac{1}{n} \sum_{i=1}^{r} \sum_{j=1}^{n_i} X_{ij} = \frac{1}{n} \sum_{i=1}^{r} n_i \bar{X}_i,
$$
其中$n = \sum_{i=1}^{r} n_i$.

计算各样本值与总的平均值之间的离差平方和：
$$
\begin{aligned}
S_T &= \sum_{i=1}^{r} \sum_{j=1}^{n_i} (X_{ij} - \bar{X})^2 = \sum_{i=1}^{r} \sum_{j=1}^{n_i} [(X_{ij} - \bar{X}_i) + (\bar{X}_i - \bar{X})]^2 \\
&= \sum_{i=1}^{r} \sum_{j=1}^{n_i} (X_{ij} - \bar{X}_i)^2 + 2 \sum_{i=1}^{r} \sum_{j=1}^{n_i} (X_{ij} - \bar{X}_i)(\bar{X}_i - \bar{X}) + \sum_{i=1}^{r} \sum_{j=1}^{n_i} (\bar{X}_i - \bar{X})^2 \\
&= \sum_{i=1}^{r} \sum_{j=1}^{n_i} (X_{ij} - \bar{X}_i)^2 + \sum_{i=1}^{r} n_i (\bar{X}_i - \bar{X})^2,
\end{aligned}
$$
记
$$
S_E = \sum_{i=1}^{r} \sum_{j=1}^{n_i} (X_{ij} - \bar{X}_i)^2,\quad S_A = \sum_{i=1}^{r} n_i (\bar{X}_i - \bar{X})^2,
$$
则有
$$
S_T = S_E + S_A,
$$
该式称为离差平方和分解公式，其中$S_E$反映了各组内部$X_{ij}$之间的差异 (即同一组内随机抽样产生的误差)，称为组内离差平方和；$S_A$反映各组之间由于因素水平不同而引起的差异 (不同水平下的差异即条件误差)，称为组间离差平方和. 方差分析就是通过对组内、组间离差平方和的比较来检验假设的.

在实际中，为了简化计算步骤，常按下面一组公式去计算$S_T$, $S_E$, $S_A$.
$$
S_T = \sum_{i=1}^{r} \sum_{j=1}^{n_i} X_{ij}^2 - \frac{1}{n} \left( \sum_{i=1}^{r} \sum_{j=1}^{n_i} X_{ij} \right)^2,
$$
$$
S_E = \sum_{i=1}^{r} \sum_{j=1}^{n_i} X_{ij}^2 - \sum_{i=1}^{r} \frac{1}{n_i} \left( \sum_{j=1}^{n_i} X_{ij} \right)^2,
$$

$$
S_A = S_T - S_E.
$$

定理 4.1
(1) $E(S_A) = (r - 1)\sigma^2 + \sum_{i=1}^{r} n_i (\mu_i - \mu)^2$;

(2) $E(S_E) = (n - r)\sigma^2$;

(3) $\frac{S_E}{\sigma^2} \sim \chi^2(n - r)$;

其中，$\mu = \frac{1}{n} \sum_{i=1}^{r} n_i \mu_i$.

定理 4.2 当$H_0: \mu_1 = \mu_2 = \cdots = \mu_r$成立时，有

(1) $\frac{S_A}{\sigma^2} \sim \chi^2(r - 1)$;

(2) $S_E$与$S_A$相互独立且$F = \frac{S_A/(r - 1)}{S_E/(n - r)} \sim F(r - 1, n - r)$.

根据定理 4.2，构造检验统计量
$$
F = \frac{S_A/(r - 1)}{S_E/(n - r)},
$$
当$H_0$成立时，$F \sim F(r - 1, n - r)$. 于是，在给定的显著性水平$\alpha$下，检验假设
$$
H_0: \mu_1 = \mu_2 = \cdots = \mu_r,
$$
若由试验数据算得结果有$F > F_{\alpha}(r - 1, n - r)$，则拒绝$H_0$，即认为因素$A$对试验结果有显著影响；若$F < F_{\alpha}(r - 1, n - r)$，则接受$H_0$，即认为因素$A$对试验结果没有显著影响.

在方差分析中，还作如下规定：

(1) 如果取$\alpha = 0.01$时拒绝$H_0$，即$F > F_{0.01}(r - 1, n - r)$，则称因素$A$的影响高度显著.

(2) 如果取$\alpha = 0.05$时拒绝$H_0$，但取$\alpha = 0.01$时不拒绝$H_0$，即
$$
F_{0.01}(r - 1, n - r) \geq F > F_{0.05}(r - 1, n - r),
$$
则称因素$A$的影响显著.

将上述统计过程归纳为方差分析表，见表 4.12. 最后一列给出$F(r - 1, n - r)$分布大于$F$值的概率$p$，当$p < \alpha$时拒绝原假设，否则接受原假设.

\begin{center}
\begin{tabular}{c|c|c|c|c|c}
\hline
\multicolumn{6}{c}{表 4.12 单因素方差分析表} \\
\hline
方差来源 & 平方和 & 自由度 & 均方和 & $F$值 & 概率 \\
\hline
因素$A$(组间) & $S_A$ & $r - 1$ & $S_A/(r - 1)$ & $F = \frac{S_A/(r - 1)}{S_E/(n - r)}$ & $p$ \\
误差(组内) & $S_E$ & $n - r$ & $S_E/(n - r)$ &  &  \\
总和 & $S_T$ & $n - 1$ &  &  &  \\
\hline
\end{tabular}
\end{center}


下面使用 statsmodels 库中的 anova_lm 函数进行单因素方差分析，输出值 F 是 $F$ 统计量的值， 输出值PR是一个概率值， 当 $PR > \alpha$($\alpha$ 为显著性水平）时接受原假设，即认为因素 A对指标无显著影响

#### sm.formula.ols(formula, data, subset=None, drop_cols=None)

OLS(Ordinary Least Squares普通最小二乘法)

| 参数          | 类型               | 说明                                    |
| ----------- | ---------------- | ------------------------------------- |
| `formula`   | str              | 形如 `"y ~ x1 + C(factor)"`，遵循 patsy 语法 |
| `data`      | DataFrame 或 dict | 变量名对应的列数据                             |
| `subset`    | array-like       | 可选，只使用部分行                             |
| `drop_cols` | list             | 可选，强制丢弃某些列                            |

###### 返回值：未拟合的 OLS 实例，需要再 .fit()

#### sm.stats.anova_lm(model, typ=2, robust=None)

anova_lm(analysis of variance linear model方差分析线性模型)

作用：对已拟合的线性模型（OLS、GLM、RLM 等）生成或补充 ANOVA 表

| 参数       | 类型                | 说明                                                |
| -------- | ----------------- | ------------------------------------------------- |
| `model`  | 已拟合结果对象           | 必传，通常是 `OLS.fit()` 返回的 `RegressionResultsWrapper` |
| `typ`    | 1, 2, 或 3         | 指定平方和类型（Type I/II/III）。绝大多数场景用 2 或 3。             |
| `robust` | `None`, `"hc0"` 等 | 是否使用异方差稳健协方差校正。默认 `None` 即可。                      |

返回值：pandas.DataFrame 包含列  
sum_sq（平方和）  
df（自由度）  
mean_sq（均方）  
F（F 值）  
PR(>F)（p 值）  

In [7]:
import numpy as np
import statsmodels.api as sm
y = np.array([1620, 1670, 1700, 1750, 1800, 1580, 1600, 1640, 1720, 1460, 1540, 1620, 1680, 1500, 1550, 1610])
x = np.hstack([np.ones(5), np.full(4, 2), np.full(4, 3), np.full(3, 4)])  # y和x shape相同
d = {'x': x, 'y': y}  # 构造字典
model = sm.formula.ols("y~C(x)", d).fit()  # 构建模型，返回RegressionResultsWrapper对象
anovat = sm.stats.anova_lm(model)  # 进行单因素分析，返回dataframe对象
anovat

,df,sum_sq,mean_sq,F,PR(>F)
C(x),3.0,60153.333333,20051.111111,3.727742,0.042004
Residual,12.0,64546.666667,5378.888889,NaN,NaN


numpy.full(shape, fill_value, dtype=None, order='K')

| 参数           | 类型          | 说明                    |
| ------------ | ----------- | --------------------- |
| `shape`      | int 或 tuple | 目标形状，例如 `4` 或 `(2,3)` |
| `fill_value` | 标量或数组       | 要填充的常量值               |
| `dtype`      | 数据类型，可选     | 默认由 `fill_value` 推断   |


In [15]:
import numpy as np
import pandas as pd
import statsmodels.api as sm
df = pd.read_excel("Pdata4_23.xlsx", header=None)
# .values转换成数组，然后转置，铺平
a = df.to_numpy().T.flatten()
b = np.arange(1, 6)  # （5,）
# tile重复，将b当成单个元素组成4x1数组
x = np.tile(b, (4, 1)).T.flatten()
d = {'x': x, 'y': a}  # 构造求解需要的字典
model = sm.formula.ols("y~C(x)", d).fit()  # 构建模型
anovat = sm.stats.anova_lm(model)
anovat

,df,sum_sq,mean_sq,F,PR(>F)
C(x),4.0,6125.7,1531.425,2.261741,0.110913
Residual,15.0,10156.5,677.100,NaN,NaN


求得 $PR= 0.110913 > \alpha = 0.05$，故认为 5 名工人的生产率没有显著差异．  
$\textbf{注 4.3}$ 认为 5 名工人的劳动生产率没有显著差异，是将 5 名工人的生产率作为一个整体进行假设检验的结果，并不表明取其中 2 名工人的生产率作两总体的均值检验时，也一定接受均值相等的假设.

### 双因素方差分析（Two-way analysis of variance）

如果要考虑两个因素对指标的影响，就要采用双因素方差分析．它的基本思想是：对每个因素各取几个水平，然后对各因素不同水平的每个组合作一次或若干次试验，对所得数据进行方差分析．对双因素方差分析可分为无重复和等重复试验两种情况，无重复试验只需检验两因素是否分别对指标有显著影响；而对等重复试验还要进一步检验两因素是否对指标有显著的交互影响.

##### 1. 数学模型

设$A$取$s$个水平$A_1, A_2, \cdots, A_s$，$B$取$r$个水平$B_1, B_2, \cdots, B_r$，在水平组合$(B_i, A_j)$下总体$X_{ij}$服从正态分布$N(\mu_{ij}, \sigma^2)$，$i = 1, \cdots, r$, $j = 1, \cdots, s$. 又设在水平组合$(B_i, A_j)$下作了$t$个试验，所得结果记作$X_{ijk}$，$X_{ijk}$服从$N(\mu_{ij}, \sigma^2)$，$i = 1, \cdots, r$, $j = 1, \cdots, s$, $k = 1, \cdots, t$，且相互独立. 将这些数据列成表 4.14 的形式.

\begin{center}
\begin{tabular}{c|c|c|c}
\hline
\multicolumn{4}{c}{表 4.14 双因素试验数据表} \\
\hline
 & $A_1$ & $A_2$ & $A_s$ \\
\hline
$B_1$ & $X_{111}, \cdots, X_{11t}$ & $X_{121}, \cdots, X_{12t}$ & $X_{1s1}, \cdots, X_{1st}$ \\
\hline
$B_2$ & $X_{211}, \cdots, X_{21t}$ & $X_{221}, \cdots, X_{22t}$ & $X_{2s1}, \cdots, X_{2st}$ \\
\hline
$\vdots$ & $\vdots$ & $\vdots$ & $\vdots$ \\
\hline
$B_r$ & $X_{r11}, \cdots, X_{r1t}$ & $X_{r21}, \cdots, X_{r2t}$ & $X_{rs1}, \cdots, X_{rst}$ \\
\hline
\end{tabular}
\end{center}

将$X_{ijk}$分解为
$$
X_{ijk} = \mu_{ij} + \varepsilon_{ijk},\quad i = 1, \cdots, r,\quad j = 1, \cdots, s,\quad k = 1, \cdots, t,
$$
其中$\varepsilon_{ijk} \sim N(0, \sigma^2)$，且相互独立. 记
$$
\mu = \frac{1}{rs} \sum_{i=1}^{r} \sum_{j=1}^{s} \mu_{ij},\quad \mu_{\cdot j} = \frac{1}{r} \sum_{i=1}^{r} \mu_{ij},\quad \alpha_j = \mu_{\cdot j} - \mu,
$$
$$
\mu_i = \frac{1}{s} \sum_{j=1}^{s} \mu_{ij},\quad \beta_i = \mu_i - \mu,\quad \gamma_{ij} = (\mu_{ij} - \mu) - \alpha_j - \beta_i,
$$
$\mu$是总均值，$\alpha_j$是水平$A_j$对指标的效应，$\beta_i$是水平$B_i$对指标的效应，$\gamma_{ij}$是水平$B_i$与$A_j$对指标的交互效应. 模型表为
$$
\left\{
\begin{aligned}
& X_{ijk} = \mu + \alpha_j + \beta_i + \gamma_{ij} + \varepsilon_{ijk}, \\
& \sum_{j=1}^{s} \alpha_j = 0,\ \sum_{i=1}^{r} \beta_i = 0,\ \sum_{i=1}^{r} \gamma_{ij} = \sum_{j=1}^{s} \gamma_{ij} = 0, \\
& \varepsilon_{ijk} \sim N(0, \sigma^2),\quad i = 1, \cdots, r;\ j = 1, \cdots, s;\ k = 1, \cdots, t.
\end{aligned}
\right. 
$$

原假设为
$$
H_{01}: \alpha_j = 0\ (j = 1, \cdots, s), 
$$
$$
H_{02}: \beta_i = 0\ (i = 1, \cdots, r), 
$$
$$
H_{03}: \gamma_{ij} = 0\ (i = 1, \cdots, r;\ j = 1, \cdots, s). \tag{4.17}
$$



例 4.24 表 4.17 给出了某种化工过程在三种浓度、四种温度水平下的得率数据. 试在水平$\alpha = 0.05$下，检验在不同温度 (因素$A$)、不同浓度 (因素$B$) 下的得率是否有显著差异？交互作用是否显著？

| **浓度 (因素$B$)** | **温度 (因素$A$)** | 10       | 24       | 38       | 52       |
|--------------------|--------------------|----------|----------|----------|----------|
| **2**              |                    | **11**   | **11**   | **13**   | **10**   |
|                    |                    | 10       | 11       | 9        | 12       |
| **4**              |                    | **9**    | **10**   | **7**    | **6**    |
|                    |                    | 7        | 8        | 11       | 10       |
| **6**              |                    | **5**    | **13**   | **12**   | **14**   |
|                    |                    | 11       | 14       | 13       | 10       |

**表 4.17 某化工过程在三种浓度、四种温度下的得率**

表 4.17 某化工过程在三种浓度、四种温度下的得率

In [31]:
import numpy as np
import statsmodels.api as sm
y = np.array([[11, 11, 13, 10], [10, 11, 9, 12], [9, 10, 7, 6], [7, 8, 11, 10], [5, 13, 12, 14], [11, 14, 13, 10]]).flatten()
A = np.tile(np.arange(1, 5), (6, 1)).flatten()
B = np.tile(np.arange(1, 4).reshape(3, 1), (1, 8)).flatten()
d = {'x1': A, 'x2': B, 'y': y}
# C(x1):A 的效应；C(x2):B 的效应;C(x1):C(x2):A 和 B 的交互作用
model = sm.formula.ols("y~C(x1)+C(x2)+C(x1):C(x2)", d).fit()
anovat = sm.stats.anova_lm(model)
anovat

,df,sum_sq,mean_sq,F,PR(>F)
C(x1),3.0,19.125000,6.375000,1.330435,0.310404
C(x2),2.0,40.083333,20.041667,4.182609,0.041856
C(x1):C(x2),6.0,18.250000,3.041667,0.634783,0.701009
Residual,12.0,57.500000,4.791667,NaN,NaN


其中 PR = 0.310404, 0.041856, 0.701009, 即认为温度影响不显著， 而浓度因素有显著差异， 交互作用不显著

注4.4 交互作用公式C(x1)+C(x2)+C(x1):C(x2) 也可以缩写为 C(xl)*C(x2),
其中大写的 C表示分类变量．

| 公式片段          | 数学意义                     |
| ------------- | ------------------------ |
| `C(x1)`       | 4 个水平的 **x1 主效应**（α₁…α₄） |
| `C(x2)`       | 3 个水平的 **x2 主效应**（β₁…β₃） |
| `C(x1):C(x2)` | **x1×x2 交互效应**（γ₁₁…γ₄₃）  |


因为 不含截距 会欠定，statsmodels 默认 自动加截距，所以实际设计矩阵为：

$$
\left[
1, \underbrace{\text{dummy}(x1)}_{\text{3列}}, \underbrace{\text{dummy}(x2)}_{\text{2列}}, \underbrace{\text{dummy}(x1 \times x2)}_{\text{6列}}
\right]_{24 \times 12}
$$

自由度：  
截距 1  
x1 主效应 3 (4−1)  
x2 主效应 2 (3−1)  
交互 6 (3×2)  
误差 12 (24−1−3−2−6)